### ISSUE TITLE
Convert APT docs to Markdown?

### ISSUE THREAD

**Issue by aadrian** [super-csv/super-csv#126]
Hi,

APT seems quite an abandoned format. Are there any plans moving to Markdown, so that more people can contribute to the docs?

Thank you.

**Comment by kbilsted** [comment]
The documentation was driven by a former member who is no longer active on the project. 

Before any conversion, what are the current struggles you are facing with updating the documentation?

**Comment by aadrian** [comment]
> Before any conversion, what are the current struggles you are facing with updating the documentation?
 
 - missing of a decent editor
 - not possible to edit and preview (mostly for small stuff) using the GitHub UI
 - extra step to generate a PDF
 - nobody is using that format anymore, whereas Markdown seems to be used everywhere these days.

just to mention a few.

**Comment by kbilsted** [comment]
Do you want to be in charge of the conversion?

**Comment by aadrian** [comment]
> Do you want to be in charge of the conversion?

Yes.

**Comment by kbilsted** [comment]
Sounds great!! :-)

Before undertaking the task, consider if there are any processes such as relase notes or referencing code files or other things in the existing documentation format that will be more error prone or manual with the conversion? Like does the compilation step of the documentation guarantee valid links and consistency?

**Comment by aadrian** [comment]
> Like does the compilation step of the documentation guarantee valid links and consistency?

I'm not aware of such a thing. If the source file (markdown or whatever format) contains a 404 link, the output will also contain a 404 link. 

There are however tools to check if all links work, but I'm not aware of any working with Maven.

**Comment by kbilsted** [comment]
how are you planning on converting the menus and stuff?

**Comment by aadrian** [comment]
> how are you planning on converting the menus and stuff?

Maven supports for building a site not just only APT, but many other formats too, Markdown being only one of them.

https://maven.apache.org/doxia/references/index.html

**Comment by kbilsted** [comment]
great

**Comment by aadrian** [comment]
@kbilsted what about not (yet?) merged pulled requests? They seem to contain have many *.apt changes too.

**Comment by kbilsted** [comment]
That is a very good point. The problem is that at the moment the project is near a stand-still. People signing up for helping out quickly fade away. I have very limited resources myself to do more than just answering tickets and occasionally reviewing code

**Comment by aadrian** [comment]
> People signing up for helping out quickly fade away.

I understand. But what is the status of those pull reuqests? Can't they simply be merged? Or are they going to be rejected? (if the branches start to go apart, than mering results in conflicts with time, or other commits make them generate conflicts)

> I have very limited resources myself to do more than just answering tickets and occasionally reviewing code

I could try to help with code (e.g. starting with #123 and #124 ), but what I don't understand yet is what's considered by the Super-CSV Team an acceptable API and functionality addition.

**Comment by ash211** [comment]
On the point that contributors fade away, I could be considered one of those people so want to provide my perspective.  I think most people drop off (like myself) because even if you get a PR merged, it takes quite some time to see your contribution in a release.

The timeline for my interactions with this project were:
- May 2016 - comment on a bug that I'd appreciate a fix for it - https://github.com/super-csv/super-csv/issues/14#issuecomment-221108045
- May 2016 - send in a PR to fix said issue -- https://github.com/super-csv/super-csv/pull/103
- June 2016 - PR is merged
- Sep 2016 - contribute a docs fix (last remaining release blocker) -- https://github.com/super-csv/super-csv/pull/111
- Oct 2016 - release process for 2.5.0 starts (no outstanding bugs remaining) -- https://github.com/super-csv/super-csv/pull/111
- Feb 2017 - another request for a 2.5.0 release -- https://github.com/super-csv/super-csv/issues/125
- Mar 2017 -- still no 2.5.0 release

As an outside contributor, it's now been 9mo since I sent in a PR that fixed a bug I cared about, and I still haven't seen a release that includes the bug.  I even tried to help push the release forward by writing docs for things I didn't directly have a stake in.

Since then I've begun considering this project abandoned and have started moving my code to other options (spark-csv mostly).

This is not to say that this is a bad project, the code has actually served my needs for several years.  But given that long delay between contribution and release, I can't recommend taking a new dependency here on super-csv.

The single best thing this project could do to increase community engagement is cut the 2.5.0 release to provide evidence that contributing to this project will actually benefit the community.

**Comment by aadrian** [comment]
> On the point that contributors fade away ....
>  I think most people drop off (like myself) because even if you get a PR merged, it takes quite some time to see your contribution in a release.
> ...
> As an outside contributor, it's now been 9mo since I sent in a PR that fixed a bug I cared about, and I still haven't seen a release that includes the bug. 

@ash211 
Could the issue be that active contributors don't get the rights to directly commit to the Git Repo? and also to the Sonatype group to be able to push releases ? (and this way the overhead is too much for those who still have the rights ? )

Pull Requests are not for active and continuous contributions.

Backwards compatibility is very important since so many projects depend on a given library, but this can be ensured by other means.

**Comment by kbilsted** [comment]
@ash211  you are correct that at the moment PR's are dealt with in a too slow pace. But at this point in my life I can't afford the time necesarry

**Comment by aadrian** [comment]
> But at this point in my life I can't afford the time necesarry

@kbilsted can you add me (and other users that request) to this GitHub Org, so that we can help?

**Comment by jamesbassett** [comment]
It's interesting to see how far things have come! Kasper and I don't even write Java these days - he went on to C++ and I'm using Kotlin and haven't looked back since :)

When I joined, the project was on Sourceforge (now a ghost town), built using ant (long since superceded by Maven then Gradle), used subversion, had bespoke documentation generation, and was only available for download from SF. See https://github.com/super-csv/super-csv/tree/v1.52.   I moved the project to Github, converted to git, converted to Maven, updated the documentation to use the maven site plugin (the standard at the time - only apt and fml were valid formats back then!) with hosting on github pages, added simple CI (Travis), and publishing to Maven Central (which is actually quite complicated).

While all of that breathed some new life into the project (especially Github and its PR process), as in inevitable in OSS projects, people move on or lack time/motivation and the project starts to stagnate.

The project could definitely benefit from the following:

* Update to gradle (lower barrier to entry - you just need Java installed)
* Update documentation to markdown (everyone's familiar with it these days)
* Update to use jcenter instead of maven central (it's sooo much easier to publish to)
* Update to perform releases via Github (i.e. creating a new tag in Github should trigger a release via Travis). I've used Netflix's nebula release plugin for this and it works quite well.

The biggest risk to the project is https://github.com/uniVocity/univocity-parsers which has corporate backing. It's sad, but I suspect this nice little project will eventually get forgotten. I'll always remember this project with fondness - it was my first foray into OSS development and I learnt so many things while working on it :)

**Comment by aadrian** [comment]
@jamesbassett 
> people move on or lack time/motivation and the project starts to stagnate.

No, not that's the real problem: developers submit PRs and they're not reviewed (nor rejected) - just **ignored**,  but the worst: the users don't even get commit rights so that they can do that part themselves :) .

**Comment by jamesbassett** [comment]
I understand your frustration, but I'm not aware of any OSS projects that give commit rights to anyone but the core team of developers. The usual process (and what happened when I joined) is to simply ask to be a member of the team. Usually that involves proving yourself by getting involved in PRs and answering questions other users have, but in the case of a truly dormant project it might just happen because the founder is happy to hand the reigns over to keep the project alive.

I believe Kasper was hoping @ZioberMichal would take over - I know he was super keen initially but has obviously moved on. I know @ash211 was keen to contribute (via PRs), but with Michal gone and Kasper taking the back seat his PRs went unattended and he moved on as well. When I asked to join the project was in a similar state - there were lots of feature requests and bugs but nobody to address them. It was fun to have so much to do!

Giving everyone commit access is a recipe for disaster - as a committer you understand the whole project, its history, and why things were done the way they were. It's easy to submit a PR that fixes your immediate problem, but often you don't see the big picture and your PR will require tweaking or even a different approach (which is a good thing). An open source project lives and dies by it's contributors/users, but it requires someone at the helm to steward it. I'm sure Kasper would be more than happy to give someone the reigns to keep the project alive...

**Comment by aadrian** [comment]
@jamesbassett 

> I'm not aware of any OSS projects that give commit rights to anyone but the core team of developers.

We are not talking about the Apache Foundation here, but a small project with a few KB of produced artifacts :) .

**Comment by kbilsted** [comment]
@aadrian its been a while, :-)  are you still interested?
